
<!DOCTYPE html>

<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width,initial-scale=1">
  <meta http-equiv="x-ua-compatible" content="ie=edge">
  <meta name="lang:clipboard.copy" content="Copy to clipboard">
  <meta name="lang:clipboard.copied" content="Copied to clipboard">
  <meta name="lang:search.language" content="en">
  <meta name="lang:search.pipeline.stopwords" content="True">
  <meta name="lang:search.pipeline.trimmer" content="True">
  <meta name="lang:search.result.none" content="No matching documents">
  <meta name="lang:search.result.one" content="1 matching document">
  <meta name="lang:search.result.other" content="# matching documents">
  <meta name="lang:search.tokenizer" content="[\s\-]+">

  
    <link href="https://fonts.gstatic.com/" rel="preconnect" crossorigin>
    <link href="https://fonts.googleapis.com/css?family=Roboto+Mono:400,500,700|Roboto:300,400,400i,700&display=fallback" rel="stylesheet">

    <style>
      body,
      input {
        font-family: "Roboto", "Helvetica Neue", Helvetica, Arial, sans-serif
      }

      code,
      kbd,
      pre {
        font-family: "Roboto Mono", "Courier New", Courier, monospace
      }
    </style>
  

  <link rel="stylesheet" href="../_static/stylesheets/application.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-palette.css"/>
  <link rel="stylesheet" href="../_static/stylesheets/application-fixes.css"/>
  
  <link rel="stylesheet" href="../_static/fonts/material-icons.css"/>
  
  <meta name="theme-color" content="#00020a">
  <script src="../_static/javascripts/modernizr.js"></script>
  
  
  
    <title>CCP4i2 — Model data: Sequences, alignments and coordinates &#8212; CCP4i2  documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/material.css" />
    <link rel="stylesheet" type="text/css" href="../_static/tabs.css" />
    <link rel="stylesheet" type="text/css" href="../_static/contentui.css" />
    <link rel="stylesheet" type="text/css" href="https://github.com/bwithd/sphinx-material/blob/main/sphinx_material/sphinx_material/static/stylesheets/application.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/sphinx_highlight.js"></script>
    <script src="../_static/contentui.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  
   

  </head>
  <body dir=ltr
        data-md-color-primary=#005BBB data-md-color-accent=yellow>
  
  <svg class="md-svg">
    <defs data-children-count="0">
      
      <svg xmlns="http://www.w3.org/2000/svg" width="500" height="500" viewBox="0 0 500 500" id="__gitlab"><path fill="currentColor" d="M93.667 473.347l90.684-279.097H2.983l90.684 279.097z" transform="translate(156.198 1.16)"/><path fill="currentColor" d="M221.333 473.345L130.649 194.25H3.557l217.776 279.095z" transform="translate(28.531 1.16)" opacity=".7"/><path fill="currentColor" d="M32 195.155L4.441 279.97a18.773 18.773 0 0 0 6.821 20.99l238.514 173.29L32 195.155z" transform="translate(.089 .256)" opacity=".5"/><path fill="currentColor" d="M2.667-84.844h127.092L75.14-252.942c-2.811-8.649-15.047-8.649-17.856 0L2.667-84.844z" transform="translate(29.422 280.256)"/><path fill="currentColor" d="M2.667 473.345L93.351 194.25h127.092L2.667 473.345z" transform="translate(247.198 1.16)" opacity=".7"/><path fill="currentColor" d="M221.334 195.155l27.559 84.815a18.772 18.772 0 0 1-6.821 20.99L3.557 474.25l217.777-279.095z" transform="translate(246.307 .256)" opacity=".5"/><path fill="currentColor" d="M130.667-84.844H3.575l54.618-168.098c2.811-8.649 15.047-8.649 17.856 0l54.618 168.098z" transform="translate(336.974 280.256)"/></svg>
      
    </defs>
  </svg>
  
  <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer">
  <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search">
  <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
  <a href="#general/model_data" tabindex="1" class="md-skip"> Skip to content </a>
  <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex navheader">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../index.html" title="CCP4i2  documentation"
           class="md-header-nav__button md-logo">
          
            &nbsp;
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          <span class="md-header-nav__topic">CCP4i2  documentation</span>
          <span class="md-header-nav__topic"> CCP4i2 — Model data: Sequences, alignments and coordinates </span>
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
        
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" action="../search.html" method="get" name="search">
      <input type="text" class="md-search__input" name="q" placeholder="Search"
             autocapitalize="off" autocomplete="off" spellcheck="false"
             data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>

      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            <a href="https://gitlab.com/ccp4i2/rstdocs" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__gitlab" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    
  </div>
</a>
          </div>
        </div>
      
      
  
  <script src="../_static/javascripts/version_dropdown.js"></script>
  <script>
    var json_loc = "../"versions.json"",
        target_loc = "../../",
        text = "Versions";
    $( document ).ready( add_version_dropdown(json_loc, target_loc, text));
  </script>
  

    </div>
  </nav>
</header>

  
  <div class="md-container">
    
    
    
  <nav class="md-tabs" data-md-component="tabs">
    <div class="md-tabs__inner md-grid">
      <ul class="md-tabs__list">
          <li class="md-tabs__item"><a href="../index.html" class="md-tabs__link">CCP4i2  documentation</a></li>
      </ul>
    </div>
  </nav>
    <main class="md-main">
      <div class="md-main__inner md-grid" data-md-component="container">
        
          <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="../index.html" title="CCP4i2 documentation" class="md-nav__button md-logo">
      
        <img src="../_static/" alt=" logo" width="48" height="48">
      
    </a>
    <a href="../index.html"
       title="CCP4i2 documentation">CCP4i2  documentation</a>
  </label>
    <div class="md-nav__source">
      <a href="https://gitlab.com/ccp4i2/rstdocs" title="Go to repository" class="md-source" data-md-source="github">

    <div class="md-source__icon">
      <svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 24 24" width="28" height="28">
        <use xlink:href="#__gitlab" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    
  </div>
</a>
    </div>
  
  

</nav>
              </div>
            </div>
          </div>
          <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
            <div class="md-sidebar__scrollwrap">
              <div class="md-sidebar__inner">
                
<nav class="md-nav md-nav--secondary">
    <label class="md-nav__title" for="__toc">Contents</label>
  <ul class="md-nav__list" data-md-scrollfix="">
        <li class="md-nav__item"><a href="#general-model-data--page-root" class="md-nav__link">CCP4i2 — Model data: Sequences, alignments and coordinates</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#au-contents-sequences" class="md-nav__link">AU contents - Sequences</a>
        </li>
        <li class="md-nav__item"><a href="#sequence-file-formats" class="md-nav__link">Sequence file formats</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#fasta-format" class="md-nav__link">FASTA format</a>
        </li>
        <li class="md-nav__item"><a href="#pir-format" class="md-nav__link">PIR format</a>
        </li>
        <li class="md-nav__item"><a href="#uniprot-xml-format" class="md-nav__link">Uniprot XML format</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#sequence-alignments-in-ccp4i2" class="md-nav__link">Sequence alignments in CCP4i2</a>
        </li>
        <li class="md-nav__item"><a href="#alignment-file-formats" class="md-nav__link">Alignment file formats</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#fasta-format-1" class="md-nav__link">FASTA format</a>
        </li>
        <li class="md-nav__item"><a href="#clustal-format" class="md-nav__link">Clustal format</a>
        </li>
        <li class="md-nav__item"><a href="#stockholm-format" class="md-nav__link">Stockholm format</a>
        </li>
        <li class="md-nav__item"><a href="#phylip-format" class="md-nav__link">Phylip format</a>
        </li>
        <li class="md-nav__item"><a href="#blast-format" class="md-nav__link">Blast format</a>
        </li>
        <li class="md-nav__item"><a href="#hhpred-format" class="md-nav__link">HHPred format</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#atomic-coordinate-file-formats" class="md-nav__link">Atomic coordinate file formats</a>
        </li></ul>
            </nav>
        </li>
        <li class="md-nav__item"><a href="#pdb-format" class="md-nav__link">PDB format</a>
        </li>
        <li class="md-nav__item"><a href="#mmcif-format" class="md-nav__link">MMCIF format</a><nav class="md-nav">
              <ul class="md-nav__list">
        <li class="md-nav__item"><a href="#ensemble-coordinate-files" class="md-nav__link">Ensemble coordinate files</a>
        </li>
        <li class="md-nav__item"><a href="#ligand-geometry" class="md-nav__link">Ligand geometry</a>
        </li>
        <li class="md-nav__item"><a href="#mol-files" class="md-nav__link">MOL files</a>
        </li>
        <li class="md-nav__item"><a href="#smiles-strings" class="md-nav__link">Smiles strings</a>
        </li>
        <li class="md-nav__item"><a href="#tls-files" class="md-nav__link">TLS files</a>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li></ul>
            </nav>
        </li>
    
<li class="md-nav__item"><a class="md-nav__extra_link" href="../_sources/general/model_data.rst.txt">Show Source</a> </li>

<li id="searchbox" class="md-nav__item"></li>

  </ul>
</nav>
              </div>
            </div>
          </div>
        
        <div class="md-content">
          <article class="md-content__inner md-typeset" role="main">
            
  <section id="ccp4i2-model-data-sequences-alignments-and-coordinates">
<h1 id="general-model-data--page-root">CCP4i2 — Model data: Sequences, alignments and coordinates<a class="headerlink" href="#general-model-data--page-root" title="Permalink to this heading">¶</a></h1>
<p><a class="reference external" href="CCP4i1Projects.html">CCP4i projects</a></p>
<p><a class="reference external" href="data_files.html">Experimental data</a></p>
<p><a class="reference external" href="model_data.html">Model data</a></p>
<p><a class="reference external" href="atom_selection.html">Atom selection</a></p>
<p><a class="reference external" href="servers.html">Servers</a></p>
<div class="content docutils container">
<p class="rubric" id="contents">Contents</p>
<p><a class="reference external" href="#sequences">AU contents - Sequences</a>
<a class="reference external" href="#sequence_files">Sequence file formats</a>
<a class="reference external" href="#alignments">Sequence alignments in CCP4i2</a>
<a class="reference external" href="#alignment_files">Alignment file formats</a>
<a class="reference external" href="#coordinate_files">Coordinate file formats</a>
<a class="reference external" href="#ensemble_coordinate_files">Ensemble coordinate files</a>
<a class="reference external" href="#ligand_geometry">Ligand geometry</a></p>
</div>
<section id="au-contents-sequences">
<h2 id="au-contents-sequences">AU contents - Sequences<a class="headerlink" href="#au-contents-sequences" title="Permalink to this heading">¶</a></h2>
<p><img alt="Sequence icon" class="icon" src="../../../qticons/SeqDataFile.png"/> The molecular replacement and model building tools
either require or work much better if provided with the sequence of the
structure to be solved. CCP4i2 stores the sequences in a <em>AU content</em>
file and it is best to create this file using the <em>Define AU contents</em>
task in the <em>Import..</em> module. This task enables you to import sequences
from a sequence file, a PDB coordinate file or to cut-n-paste into a
sequence editor. If you subsequently want to edit the <em>AU content</em> you
can open another <em>Define AU contents</em> task, import the existing file and
then edit as you like.</p>
<p><img alt="Sequence input" src="../_images/sequence_widget.jpg"/></p>
<p>If you need to add another sequence then click the <em>Show list</em> button at
the top of the input shown above and click the ‘+’ symbol to add another
sequence to the list.</p>
<p>There are many different formats for sequence files and they are often
poorly defined. CCP4i2 will attempt ot read any file it is presented -
please check in the viewer that a file is read correctly. If a file
contains multiple sequences you will be allowed to select one or more.
If CCP4i2 has difficulty reading a sequence file please try the <a class="reference external" href="http://www.ebi.ac.uk/Tools/sfc/readseq/">EBI
readseq</a> tool to convert to
an acceptable format.</p>
<p>Where the <em>AU content</em> is input to tasks you will usually have the
option to select one or more of the chains in the file.</p>
<p><img alt="Sequence selection" src="../_images/sequence_selection.jpg"/></p>
</section>
<section id="sequence-file-formats">
<h2 id="sequence-file-formats">Sequence file formats<a class="headerlink" href="#sequence-file-formats" title="Permalink to this heading">¶</a></h2>
<section id="fasta-format">
<h3 id="fasta-format">FASTA format<a class="headerlink" href="#fasta-format" title="Permalink to this heading">¶</a></h3>
<p>The first line starts with &gt;, then a word for the name of the sequence
and the rest of the line is a description of the sequence. The remaining
lines contain the sequence itself. Blank lines, spaces or other symbols
(dashes, underscores, periods) are ignored.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span> <span class="o">&gt;</span><span class="n">rnase</span> <span class="n">This</span> <span class="ow">is</span> <span class="n">the</span> <span class="n">rnase</span> <span class="n">sequence</span>
<span class="n">DVSGTVCLSALPPEATDTLNLIASDGPFPYSQDGVVFQNRESVLPTQSYGYYHEYTVITPGARTRGTRRIICGE</span>
<span class="n">ATQEDYYTGDHYATFSLIDQTC</span>
</pre></div>
</div>
</section>
<section id="pir-format">
<h3 id="pir-format">PIR format<a class="headerlink" href="#pir-format" title="Permalink to this heading">¶</a></h3>
<p>A PIR (sometimes called NBRF) file contains one or more sequences in the
form: Line 1: “&gt;P1;” which includes a two-letter code defining the
sequence type (P1, F1, DL, DC, RL, RC, or XX) followed by the database
ID code. Line 2: text description of the sequence. Lines 3+: the
sequence, which can include white space and ‘-’ (that will be ignored)
ending with “*” character. Optionally these can be followed by more
lines describing the sequence. Example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">P1</span><span class="p">;</span> <span class="n">RNASE</span>
<span class="n">Chain</span> <span class="n">A</span> <span class="k">for</span> <span class="n">Rnase</span>
<span class="n">DVSGTVCLSALPPEATDTLNLIASDGPFPYSQDGVVFQNRESVLPTQSYGYYHEYTVITPGARTRGTRRIICGE</span>
<span class="n">ATQEDYYTGDHYATFSLIDQTC</span><span class="o">*</span>
</pre></div>
</div>
</section>
<section id="uniprot-xml-format">
<h3 id="uniprot-xml-format">Uniprot XML format<a class="headerlink" href="#uniprot-xml-format" title="Permalink to this heading">¶</a></h3>
<p>A file in this format begins like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&lt;?xml version='1.0' encoding='UTF-8'?&gt;
&lt;uniprot xmlns="http://uniprot.org/uniprot" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://uniprot.org/uniprot http://www.uniprot.org/support/docs/uniprot.xsd"&gt;
&lt;entry dataset="Swiss-Prot" created="1997-11-01" modified="2015-02-04" version="125"&gt;
&lt;accession&gt;Q02555&lt;/accession&gt;
&lt;accession&gt;D6W065&lt;/accession&gt;
&lt;accession&gt;Q04008&lt;/accession&gt;
&lt;accession&gt;Q05038&lt;/accession&gt;
</pre></div>
</div>
<p>CCP4i2 may download such files from the Uniprot website and save them in
the project download directory.</p>
<section id="sequence-alignments-in-ccp4i2">
<h4 id="sequence-alignments-in-ccp4i2">Sequence alignments in CCP4i2<a class="headerlink" href="#sequence-alignments-in-ccp4i2" title="Permalink to this heading">¶</a></h4>
<p><img alt="Alignment icon" class="icon" src="../../../qticons/SeqAlignDataFile.png"/> Sequence alignments are used in CCP4i2 to create better
molecular replacement models using the <em>Truncate search model</em> or <em>Edit
search model</em> tools. There is an <em>Align sequences with Clustalw</em> task to
align two or more sequences or an alignment from an alternative source
can be used.</p>
</section>
<section id="alignment-file-formats">
<h4 id="alignment-file-formats">Alignment file formats<a class="headerlink" href="#alignment-file-formats" title="Permalink to this heading">¶</a></h4>
<p>Aligment files contain multiple sequences formatted to indicate the
equivalent residues. This usually requires spaces or dashes in an
individual sequence where there is no equivalent residue. Alignment
files often have extra ‘consensus’ lines with symbols indicating, for
example, conserved residues.</p>
<p>Files from <em>PsiBlast</em> or <em>HHPred</em> servers such as
<a class="reference external" href="http://www.ebi.ac.uk/Tools/sss/psiblast/">http://www.ebi.ac.uk/Tools/sss/psiblast/</a> and
<a class="reference external" href="http://toolkit.tuebingen.mpg.de/hhpred">http://toolkit.tuebingen.mpg.de/hhpred</a> may contain mutliple sets of
alignments and the CCP4i2 <em>Import an alignment</em> interface allows you to
select one of those alignments to export in Clustal format.</p>
</section>
</section>
<section id="fasta-format-1">
<span id="id1"></span><h3 id="fasta-format-1">FASTA format<a class="headerlink" href="#fasta-format-1" title="Permalink to this heading">¶</a></h3>
<p>The format is the same as the sequence format described above but with
multiple sequences which may contain spaces or dashes.</p>
</section>
<section id="clustal-format">
<h3 id="clustal-format">Clustal format<a class="headerlink" href="#clustal-format" title="Permalink to this heading">¶</a></h3>
<div class="line-block">
<div class="line">The first line begins “CLUSTAL W” or “CLUSTALW” and can then have
annotation which is followed by blank lines that are ignored. There
are one or more blocks of sequence data separated by blank lines. Each
block consists of one line for each sequence followed by a line
showing the sequence conservation. The sequence line consists of the
sequence name, white space and up to 60 sequence symbols. There can be
white space and a cumulative count of residues for the sequence. The
block ends with line showing the degree of conservation for the
columns of the alignment in this block. Gaps in sequences are
represented using hyphens.</div>
<div class="line">The characters used to represent the degree of conservation are</div>
</div>
<table>
<tbody>
<tr class="row-odd"><td><p><strong>*</strong></p></td>
<td><p>all residues or nucleotides in that column are identical</p></td>
</tr>
<tr class="row-even"><td><p><strong>:</strong></p></td>
<td><p>conserved substitutions have been observed</p></td>
</tr>
<tr class="row-odd"><td><p><strong>.</strong></p></td>
<td><p>semi-conserved substitutions have been observed
no match</p></td>
</tr>
</tbody>
</table>
<div class="line-block">
<div class="line"><br/></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">CLUSTAL</span> <span class="mf">2.1</span> <span class="n">multiple</span> <span class="n">sequence</span> <span class="n">alignment</span>


<span class="n">rnase_A</span>         <span class="o">--------------------------------------------------</span><span class="n">DVSGTVCLS</span><span class="o">-</span>
<span class="mi">1</span><span class="n">vjf_A</span>          <span class="n">MKTRADLFAFFDAHGVDHKTLDHPPVFRVEEGLEIKAAMPGGHTKNLFLKDAKGQLWLIS</span>
                                                                  <span class="o">*..*</span> <span class="p">:</span> <span class="o">*</span>

<span class="n">rnase_A</span>         <span class="n">ALPPEATDT</span><span class="o">--</span><span class="n">LNLIASDGPFPYSQDGVVFQNRESVLP</span><span class="o">--</span><span class="n">TQSYGYYHEYT</span><span class="o">-----</span><span class="n">VITP</span>
<span class="mi">1</span><span class="n">vjf_A</span>          <span class="n">ALGETTIDLKKLHHVIGSGRLSFGPQEMMLETLG</span><span class="o">-</span><span class="n">VTPGSVTAFGLINDTEKRVRFVLDK</span>
                <span class="o">**</span>   <span class="p">:</span> <span class="o">*</span>   <span class="o">*</span><span class="p">:</span> <span class="p">:</span> <span class="o">..*</span> <span class="p">:</span><span class="o">.</span><span class="p">:</span><span class="o">.</span> <span class="p">:</span> <span class="p">::::</span><span class="o">.</span>   <span class="o">*</span> <span class="o">*</span>  <span class="o">.</span> <span class="p">::</span><span class="o">*</span>  <span class="p">::</span>       <span class="o">*</span><span class="p">:</span>

<span class="n">rnase_A</span>         <span class="n">GART</span><span class="o">---------</span><span class="n">RGTRRIICGEATQEDYYTG</span><span class="o">----</span><span class="n">DHYATFSLIDQTC</span>
<span class="mi">1</span><span class="n">vjf_A</span>          <span class="n">ALADSDPVNFHPLKNDATTAVSQAGLRRFLAALGVEPMIVDFAAMEVVG</span>
                <span class="o">.</span>            <span class="p">:</span><span class="o">.</span>      <span class="o">.</span><span class="p">:</span><span class="o">*</span>  <span class="o">.</span> <span class="p">:</span> <span class="p">:</span><span class="o">.</span>       <span class="o">.</span> <span class="o">*</span><span class="p">:</span> <span class="p">::</span> <span class="o">.</span>
</pre></div>
</div>
<p>Beware there may be problems reading this file if the conservation line
is shorter than the sequence lines in the same block - the conservation
line needs to have spaces on the end of the line where there is no
conservation. Also there should be NO blank lines on the end of the
file.</p>
</section>
<section id="stockholm-format">
<h3 id="stockholm-format">Stockholm format<a class="headerlink" href="#stockholm-format" title="Permalink to this heading">¶</a></h3>
<p>Also called PFAM format, is described
<a class="reference external" href="http://sonnhammer.sbc.su.se/Stockholm.html">http://sonnhammer.sbc.su.se/Stockholm.html</a></p>
</section>
<section id="phylip-format">
<h3 id="phylip-format">Phylip format<a class="headerlink" href="#phylip-format" title="Permalink to this heading">¶</a></h3>
<p>Is described
<a class="reference external" href="http://scikit-bio.org/docs/0.2.2/generated/skbio.io.phylip.html">http://scikit-bio.org/docs/0.2.2/generated/skbio.io.phylip.html</a></p>
</section>
<section id="blast-format">
<h3 id="blast-format">Blast format<a class="headerlink" href="#blast-format" title="Permalink to this heading">¶</a></h3>
<p>Blast stands for ‘Basic Local Alignment Search Tool’ - the objective of
Blast is to find <strong>regions</strong> of sequence similarity and the output file
shows only the similar regions and possibly not the full alignment of
two sequences required by the tools in CCP4i2. There are alternative
Blast output files and CCP4i2 will read only the XML file.</p>
</section>
<section id="hhpred-format">
<h3 id="hhpred-format">HHPred format<a class="headerlink" href="#hhpred-format" title="Permalink to this heading">¶</a></h3>
<p>An HHPred file contains a header followed by a list of all hits with
statistics and then an alignment for each hit.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Query</span>         <span class="n">pdb</span><span class="o">|</span><span class="mi">1</span><span class="n">dfr</span><span class="o">|</span><span class="n">A</span> <span class="n">B</span>
<span class="n">Match_columns</span> <span class="mi">159</span>
<span class="n">No_of_seqs</span>    <span class="mi">222</span> <span class="n">out</span> <span class="n">of</span> <span class="mi">1085</span>
<span class="o">...</span>
<span class="o">...</span>
 <span class="n">No</span> <span class="n">Hit</span>                             <span class="n">Prob</span> <span class="n">E</span><span class="o">-</span><span class="n">value</span> <span class="n">P</span><span class="o">-</span><span class="n">value</span>  <span class="n">Score</span>    <span class="n">SS</span> <span class="n">Cols</span> <span class="n">Query</span> <span class="n">HMM</span>  <span class="n">Template</span> <span class="n">HMM</span>
  <span class="mi">1</span> <span class="mi">1</span><span class="n">zdr_A</span> <span class="n">Dihydrofolate</span> <span class="n">reductase</span> <span class="mf">100.0</span> <span class="mf">7.4E-46</span>   <span class="mf">2E-50</span>  <span class="mf">270.7</span>  <span class="mf">21.4</span>  <span class="mi">159</span>    <span class="mi">1</span><span class="o">-</span><span class="mi">159</span>     <span class="mi">1</span><span class="o">-</span><span class="mi">161</span> <span class="p">(</span><span class="mi">164</span><span class="p">)</span>
  <span class="mi">2</span> <span class="mi">2</span><span class="n">w9h_A</span> <span class="n">DHFR</span><span class="p">,</span> <span class="n">dihydrofolate</span> <span class="n">red</span> <span class="mf">100.0</span> <span class="mf">1.5E-45</span> <span class="mf">4.1E-50</span>  <span class="mf">267.8</span>  <span class="mf">18.6</span>  <span class="mi">156</span>    <span class="mi">1</span><span class="o">-</span><span class="mi">159</span>     <span class="mi">2</span><span class="o">-</span><span class="mi">158</span> <span class="p">(</span><span class="mi">159</span><span class="p">)</span>
  <span class="mi">3</span> <span class="mi">4</span><span class="n">m7u_A</span> <span class="n">Dihydrofolate</span> <span class="n">reductase</span> <span class="mf">100.0</span>   <span class="mf">1E-44</span> <span class="mf">2.8E-49</span>  <span class="mf">267.5</span>  <span class="mf">21.3</span>  <span class="mi">159</span>    <span class="mi">1</span><span class="o">-</span><span class="mi">159</span>     <span class="mi">9</span><span class="o">-</span><span class="mi">170</span> <span class="p">(</span><span class="mi">176</span><span class="p">)</span>
  <span class="mi">4</span> <span class="mi">3</span><span class="n">ia4_A</span> <span class="n">Dihydrofolate</span> <span class="n">reductase</span> <span class="mf">100.0</span> <span class="mf">9.5E-45</span> <span class="mf">2.6E-49</span>  <span class="mf">264.3</span>  <span class="mf">19.2</span>  <span class="mi">159</span>    <span class="mi">1</span><span class="o">-</span><span class="mi">159</span>     <span class="mi">2</span><span class="o">-</span><span class="mi">161</span> <span class="p">(</span><span class="mi">162</span><span class="p">)</span>
<span class="o">...</span>
<span class="o">...</span>
<span class="n">No</span> <span class="mi">1</span>
<span class="o">&gt;</span><span class="mi">1</span><span class="n">zdr_A</span> <span class="n">Dihydrofolate</span> <span class="n">reductase</span><span class="p">;</span> <span class="n">DHFR</span><span class="p">,</span> <span class="n">NADP</span><span class="p">,</span> <span class="n">oxidoreductase</span><span class="p">;</span> <span class="mf">2.00</span><span class="n">A</span> <span class="p">{</span><span class="n">Geobacillus</span> <span class="n">stearothermophilus</span><span class="p">}</span> <span class="n">SCOP</span><span class="p">:</span> <span class="n">c</span><span class="mf">.71.1.0</span>
<span class="n">Probab</span><span class="o">=</span><span class="mf">100.00</span>  <span class="n">E</span><span class="o">-</span><span class="n">value</span><span class="o">=</span><span class="mf">7.4e-46</span>  <span class="n">Score</span><span class="o">=</span><span class="mf">270.69</span>  <span class="n">Aligned_cols</span><span class="o">=</span><span class="mi">159</span>  <span class="n">Identities</span><span class="o">=</span><span class="mi">39</span><span class="o">%</span>  <span class="n">Similarity</span><span class="o">=</span><span class="mf">0.716</span>  <span class="n">Sum_probs</span><span class="o">=</span><span class="mf">0.0</span>

<span class="n">Q</span> <span class="n">ss_pred</span>             <span class="n">CEEEEEEECCCCcEECCCCCCCCCHHHHHHHHHHhCCCEEEEchhHHHhcCCCCCCCeEEEECCCCCCC</span><span class="o">-</span><span class="n">CCeEEECCHH</span>
<span class="n">Q</span> <span class="n">pdb</span><span class="o">|</span><span class="mi">1</span><span class="n">dfr</span><span class="o">|</span><span class="n">A</span>        <span class="mi">1</span> <span class="n">MISLIAALAVDRVIGMENAMPWNLPADLAWFKRNTLDKPVIMGRHTWESIGRPLPGRKNIILSSQPGTD</span><span class="o">-</span><span class="n">DRVTWVKSVD</span>   <span class="mi">79</span> <span class="p">(</span><span class="mi">159</span><span class="p">)</span>
<span class="n">Q</span> <span class="n">Consensus</span>         <span class="mi">1</span> <span class="o">~</span><span class="n">i</span><span class="o">~~~~</span><span class="n">a</span><span class="o">~</span><span class="n">s</span><span class="o">~</span><span class="n">dG</span><span class="o">~</span><span class="n">I</span><span class="o">~~~</span><span class="n">g</span><span class="o">~~~</span><span class="n">W</span><span class="o">~~~~</span><span class="n">d</span><span class="o">~~~</span><span class="n">f</span><span class="o">~~~~~~~~</span><span class="n">ilmGr</span><span class="o">~</span><span class="n">T</span><span class="o">~~~~~~~~~~~~~</span><span class="n">iV</span><span class="o">~</span><span class="n">s</span><span class="o">~~~~~~-~~~~~~~~~~</span>   <span class="mi">79</span> <span class="p">(</span><span class="mi">159</span><span class="p">)</span>
                      <span class="o">||++++|+|+||+||.+|++||+.++|+++|++.+.++++||||+||++++||+++|++||+||+...+</span> <span class="o">++++++.|++</span>
<span class="n">T</span> <span class="n">Consensus</span>         <span class="mi">1</span> <span class="n">mi</span><span class="o">~</span><span class="n">l</span><span class="o">~~</span><span class="n">A</span><span class="o">~</span><span class="n">sldG</span><span class="o">~</span><span class="n">Ig</span><span class="o">~~</span><span class="n">g</span><span class="o">~</span><span class="n">l</span><span class="o">~</span><span class="n">W</span><span class="o">~~~~</span><span class="n">d</span><span class="o">~~~</span><span class="n">f</span><span class="o">~~~</span><span class="n">t</span><span class="o">~~~~</span><span class="n">vlmGR</span><span class="o">~</span><span class="n">T</span><span class="o">~</span><span class="n">e</span><span class="o">~~~~</span><span class="n">pl</span><span class="o">~~</span><span class="n">r</span><span class="o">~~</span><span class="n">iV</span><span class="o">~</span><span class="n">S</span><span class="o">~~~~~~~~~~~</span><span class="n">v</span><span class="o">~~~~~</span>   <span class="mi">80</span> <span class="p">(</span><span class="mi">164</span><span class="p">)</span>
<span class="n">T</span> <span class="mi">1</span><span class="n">zdr_A</span>            <span class="mi">1</span> <span class="n">MISHIVAMDENRVIGKDNRLPWHLPADLAYFKRVTMGHAIVMGRKTFEAIGRPLPGRDNVVVTGNRSFRPEGCLVLHSLE</span>   <span class="mi">80</span> <span class="p">(</span><span class="mi">164</span><span class="p">)</span>
<span class="n">T</span> <span class="n">ss_dssp</span>             <span class="n">CEEEEEEEETTCEEEBTTBCSSCCHHHHHHHHHHHTTSEEEEEHHHHHHHCSCCTTSEEEEECSCTTCCCTTCEEECSHH</span>
<span class="n">T</span> <span class="n">ss_pred</span>             <span class="n">CEEEEEEECCCCcEECCCCcccCCHHHHHHHHHHhcCCEEEEchHHhhhccccCCCCEEEEEcCCCCCCCCCEEEECCHH</span>
<span class="o">...</span>
<span class="o">...</span>
<span class="o">...</span>
</pre></div>
</div>
<section id="atomic-coordinate-file-formats">
<h4 id="atomic-coordinate-file-formats">Atomic coordinate file formats<a class="headerlink" href="#atomic-coordinate-file-formats" title="Permalink to this heading">¶</a></h4>
<p><img alt="Coordinate icon" class="icon" src="../../../qticons/PdbDataFile.png"/>An initial coordinate model is created in <em>Molecular
replacement</em> tasks or by model building tasks such as <em>Autobuild
protein</em> after <em>Experimental phasing</em>. Improved models are created by
extending and refining the structure using <em>Model building and
refinement</em> tasks.</p>
<p>There are two alternative file formats to represent macromolecular
structures. Crystallographers are slowly moving from using long standing
PDB format to using the more flexible MMCIF format. All CCP4i2 tasks can
use either format.</p>
</section>
</section>
<section id="pdb-format">
<h3 id="pdb-format">PDB format<a class="headerlink" href="#pdb-format" title="Permalink to this heading">¶</a></h3>
<p>A pdb file normally has the extension <em>.pdb</em> though files downloaded
from the structure databases have the form <em>pdb</em>xnyz.<em>ent</em>. The file
may have extensive header information, particularly if it is from a
structure database, such as:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">HEADER</span>    <span class="n">OXIDOREDUCTASE</span>                          <span class="mi">14</span><span class="o">-</span><span class="n">JUL</span><span class="o">-</span><span class="mi">10</span>   <span class="mi">3</span><span class="n">NXO</span>
<span class="n">TITLE</span>     <span class="n">PERFERENTIAL</span> <span class="n">SELECTION</span> <span class="n">OF</span> <span class="n">ISOMER</span> <span class="n">BINDING</span> <span class="n">FROM</span><span class="o">..</span>
<span class="o">..</span>
<span class="o">..</span>
<span class="n">CRYST1</span>   <span class="mf">84.150</span>   <span class="mf">84.150</span>   <span class="mf">77.968</span>  <span class="mf">90.00</span>  <span class="mf">90.00</span> <span class="mf">120.00</span> <span class="n">H</span> <span class="mi">3</span>           <span class="mi">9</span>
<span class="n">ORIGX1</span>      <span class="mf">1.000000</span>  <span class="mf">0.000000</span>  <span class="mf">0.000000</span>        <span class="mf">0.00000</span>
<span class="n">ORIGX2</span>      <span class="mf">0.000000</span>  <span class="mf">1.000000</span>  <span class="mf">0.000000</span>        <span class="mf">0.00000</span>
<span class="n">ORIGX3</span>      <span class="mf">0.000000</span>  <span class="mf">0.000000</span>  <span class="mf">1.000000</span>        <span class="mf">0.00000</span>
<span class="n">SCALE1</span>      <span class="mf">0.011884</span>  <span class="mf">0.006861</span>  <span class="mf">0.000000</span>        <span class="mf">0.00000</span>
<span class="n">SCALE2</span>      <span class="mf">0.000000</span>  <span class="mf">0.013722</span>  <span class="mf">0.000000</span>        <span class="mf">0.00000</span>
<span class="n">SCALE3</span>      <span class="mf">0.000000</span>  <span class="mf">0.000000</span>  <span class="mf">0.012826</span>        <span class="mf">0.00000</span>
</pre></div>
</div>
<p>followed by atomic coordinates in the form:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ATOM</span>      <span class="mi">1</span>  <span class="n">N</span>   <span class="n">VAL</span> <span class="n">A</span>   <span class="mi">1</span>      <span class="mf">19.401</span>  <span class="mf">29.704</span>  <span class="o">-</span><span class="mf">2.475</span>  <span class="mf">1.00</span> <span class="mf">24.77</span>           <span class="n">N</span>
<span class="n">ATOM</span>      <span class="mi">2</span>  <span class="n">CA</span>  <span class="n">VAL</span> <span class="n">A</span>   <span class="mi">1</span>      <span class="mf">19.249</span>  <span class="mf">28.215</span>  <span class="o">-</span><span class="mf">2.218</span>  <span class="mf">1.00</span> <span class="mf">21.14</span>           <span class="n">C</span>
<span class="n">ATOM</span>      <span class="mi">3</span>  <span class="n">C</span>   <span class="n">VAL</span> <span class="n">A</span>   <span class="mi">1</span>      <span class="mf">19.972</span>  <span class="mf">27.926</span>  <span class="o">-</span><span class="mf">0.913</span>  <span class="mf">1.00</span> <span class="mf">19.93</span>           <span class="n">C</span>
<span class="n">ATOM</span>      <span class="mi">4</span>  <span class="n">O</span>   <span class="n">VAL</span> <span class="n">A</span>   <span class="mi">1</span>      <span class="mf">20.957</span>  <span class="mf">28.590</span>  <span class="o">-</span><span class="mf">0.597</span>  <span class="mf">1.00</span> <span class="mf">18.52</span>           <span class="n">O</span>
<span class="n">ATOM</span>      <span class="mi">5</span>  <span class="n">CB</span>  <span class="n">VAL</span> <span class="n">A</span>   <span class="mi">1</span>      <span class="mf">19.819</span>  <span class="mf">27.388</span>  <span class="o">-</span><span class="mf">3.372</span>  <span class="mf">1.00</span> <span class="mf">23.70</span>           <span class="n">C</span>
<span class="n">ATOM</span>      <span class="mi">6</span>  <span class="n">CG1</span> <span class="n">VAL</span> <span class="n">A</span>   <span class="mi">1</span>      <span class="mf">19.767</span>  <span class="mf">25.871</span>  <span class="o">-</span><span class="mf">3.097</span>  <span class="mf">1.00</span> <span class="mf">23.27</span>           <span class="n">C</span>
<span class="n">ATOM</span>      <span class="mi">7</span>  <span class="n">CG2</span> <span class="n">VAL</span> <span class="n">A</span>   <span class="mi">1</span>      <span class="mf">19.090</span>  <span class="mf">27.701</span>  <span class="o">-</span><span class="mf">4.705</span>  <span class="mf">1.00</span> <span class="mf">24.61</span>           <span class="n">C</span>
<span class="n">ATOM</span>      <span class="mi">8</span>  <span class="n">N</span>   <span class="n">GLY</span> <span class="n">A</span>   <span class="mi">2</span>      <span class="mf">19.456</span>  <span class="mf">27.017</span>  <span class="o">-</span><span class="mf">0.098</span>  <span class="mf">1.00</span> <span class="mf">15.48</span>           <span class="n">N</span>
<span class="n">ATOM</span>      <span class="mi">9</span>  <span class="n">CA</span>  <span class="n">GLY</span> <span class="n">A</span>   <span class="mi">2</span>      <span class="mf">20.053</span>  <span class="mf">26.668</span>   <span class="mf">1.166</span>  <span class="mf">1.00</span> <span class="mf">13.83</span>           <span class="n">C</span>
<span class="n">ATOM</span>     <span class="mi">10</span>  <span class="n">C</span>   <span class="n">GLY</span> <span class="n">A</span>   <span class="mi">2</span>      <span class="mf">21.098</span>  <span class="mf">25.532</span>   <span class="mf">1.080</span>  <span class="mf">1.00</span> <span class="mf">12.09</span>           <span class="n">C</span>
<span class="n">ATOM</span>     <span class="mi">11</span>  <span class="n">O</span>   <span class="n">GLY</span> <span class="n">A</span>   <span class="mi">2</span>      <span class="mf">21.850</span>  <span class="mf">25.441</span>   <span class="mf">0.092</span>  <span class="mf">1.00</span> <span class="mf">12.97</span>           <span class="n">O</span>
<span class="n">ATOM</span>     <span class="mi">12</span>  <span class="n">N</span>   <span class="n">SER</span> <span class="n">A</span>   <span class="mi">3</span>      <span class="mf">21.141</span>  <span class="mf">24.739</span>   <span class="mf">2.150</span>  <span class="mf">1.00</span> <span class="mf">11.92</span>
<span class="n">N</span>
<span class="o">...</span>
</pre></div>
</div>
<p>The full PDB format is formally described
<a class="reference external" href="http://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html">http://www.wwpdb.org/documentation/file-format-content/format33/v3.3.html</a>.</p>
</section>
<section id="mmcif-format">
<h3 id="mmcif-format">MMCIF format<a class="headerlink" href="#mmcif-format" title="Permalink to this heading">¶</a></h3>
<p>An MMCIF file normally has the extension <em>.cif</em> but beware the format is
very flexible and a cif format file might alternatively contain
experimental data or other crystallographic data. The header of a cif
file is very variable but likely contains the cell parameters in a form:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">_cell</span><span class="o">.</span><span class="n">entry_id</span>               <span class="mi">1</span><span class="n">VDR</span>
<span class="n">_cell</span><span class="o">.</span><span class="n">length_a</span>               <span class="mf">70.870</span>
<span class="n">_cell</span><span class="o">.</span><span class="n">length_b</span>               <span class="mf">59.450</span>
<span class="n">_cell</span><span class="o">.</span><span class="n">length_c</span>               <span class="mf">78.150</span>
<span class="n">_cell</span><span class="o">.</span><span class="n">angle_alpha</span>            <span class="mf">90.00</span>
<span class="n">_cell</span><span class="o">.</span><span class="n">angle_beta</span>             <span class="mf">95.80</span>
<span class="n">_cell</span><span class="o">.</span><span class="n">angle_gamma</span>            <span class="mf">90.00</span>
</pre></div>
</div>
<p>The atomic coordinates are in the form:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>loop_
_atom_site.group_PDB
_atom_site.id
_atom_site.type_symbol
_atom_site.label_atom_id
_atom_site.label_alt_id
_atom_site.label_comp_id
_atom_site.label_asym_id
_atom_site.label_entity_id
_atom_site.label_seq_id
_atom_site.pdbx_PDB_ins_code
_atom_site.Cartn_x
_atom_site.Cartn_y
_atom_site.Cartn_z
_atom_site.occupancy
_atom_site.B_iso_or_equiv
_atom_site.Cartn_x_esd
_atom_site.Cartn_y_esd
_atom_site.Cartn_z_esd
_atom_site.occupancy_esd
_atom_site.B_iso_or_equiv_esd
_atom_site.pdbx_formal_charge
_atom_site.auth_seq_id
_atom_site.auth_comp_id
_atom_site.auth_asym_id
_atom_site.auth_atom_id
_atom_site.pdbx_PDB_model_num
_atom_site.pdbe_label_seq_id
ATOM 1 N N . GLU A 1 2 ? 0.460 -1.784 25.242 1.00 40.49 ? ? ? ? ? ? 2 GLU A N 1 2
ATOM 2 C CA . GLU A 1 2 ? 1.146 -0.578 25.666 1.00 38.88 ? ? ? ? ? ? 2 GLU A CA 1 2
ATOM 3 C C . GLU A 1 2 ? 2.671 -0.669 25.658 1.00 36.68 ? ? ? ? ? ? 2 GLU A C 1 2
ATOM 4 O O . GLU A 1 2 ? 3.268 -1.615 26.185 1.00 32.84 ? ? ? ? ? ? 2 GLU A O 1 2
ATOM 5 C CB . GLU A 1 2 ? 0.685 -0.189 27.070 1.00 41.47 ? ? ? ? ? ? 2 GLU A CB 1 2
ATOM 6 C CG . GLU A 1 2 ? 0.860 -1.234 28.155 1.00 44.68 ? ? ? ? ? ? 2 GLU A CG 1 2
ATOM 7 C CD . GLU A 1 2 ? 1.332 -0.644 29.475 1.00 49.47 ? ? ? ? ? ? 2 GLU A CD 1 2
ATOM 8 O OE1 . GLU A 1 2 ? 0.557 0.044 30.149 1.00 49.28 ? ? ? ? ? ? 2 GLU A OE1 1 2
ATOM 9 O OE2 . GLU A 1 2 ? 2.491 -0.882 29.820 1.00 51.50 ? ? ? ? ? ? 2 GLU A OE2 1 2
ATOM 10 N N . LEU A 1 3 ? 3.294 0.332 25.011 1.00 34.68 ? ? ? ? ? ? 3 LEU A N 1 3
ATOM 11 C CA . LEU A 1 3 ? 4.753 0.418 24.969 1.00 31.97 ? ? ? ? ? ? 3 LEU A CA 1 3
ATOM 12 C C . LEU A 1 3 ? 5.232 1.003 26.284 1.00 28.62 ? ? ? ? ? ? 3 LEU A C 1 3
ATOM 13 O O . LEU A 1 3 ? 4.657 1.960 26.784 1.00 26.46 ? ? ? ? ? ? 3 LEU A O 1 3
ATOM 14 C CB . LEU A 1 3 ? 5.220 1.308 23.809 1.00 32.60 ? ? ? ? ? ? 3
LEU A CB 1 3
..
</pre></div>
</div>
<p>The MMCIF format is formally described
<a class="reference external" href="http://mmcif.wwpdb.org/">http://mmcif.wwpdb.org</a></p>
<section id="ensemble-coordinate-files">
<h4 id="ensemble-coordinate-files">Ensemble coordinate files<a class="headerlink" href="#ensemble-coordinate-files" title="Permalink to this heading">¶</a></h4>
<p>Ensemble coordinate files contain multiple putative model coordinates
for a structure that are used in some MR tasks such as Phaser. The files
are in the the usual coordinate formats: PDB or mmCIf but use the
‘MODEL’ feature, normally used to support multiple NMR models, to
support multiple putative MR models.</p>
</section>
<section id="ligand-geometry">
<h4 id="ligand-geometry">Ligand geometry<a class="headerlink" href="#ligand-geometry" title="Permalink to this heading">¶</a></h4>
<p><img alt="Ligand geometry icon" class="icon" src="../../../qticons/DictDataFile.png"/> Refinement programs attempt to make the structure
geometry (bond lengths and angles etc.) conform to accepted ideal values
that are given in a geometry library ($CCP4/lib/data/monomers). The
library contains data for common macromolecular residues and ligands. If
your structure contains a novel ligand that is not in the library then
the <em>Make Ligand</em> task will make a Geometry file and a coordinate file
with an initial structure for a ligand. You can enter either a <a class="reference external" href="https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system">SMILES
string</a>
or a <a class="reference external" href="https://en.wikipedia.org/wiki/Chemical_table_file">MOL file</a> or
sketch the molecule with the <em>Lydia</em> program.</p>
</section>
<section id="mol-files">
<h4 id="mol-files">MOL files<a class="headerlink" href="#mol-files" title="Permalink to this heading">¶</a></h4>
<p><img alt="MOL file icon" class="icon" src="../../../svgicons/MDLMolDataFile.svg"/>See
<a class="reference external" href="https://en.wikipedia.org/wiki/Chemical_table_file">Wikipedia</a>.</p>
</section>
<section id="smiles-strings">
<h4 id="smiles-strings">Smiles strings<a class="headerlink" href="#smiles-strings" title="Permalink to this heading">¶</a></h4>
<p>See
<a class="reference external" href="https://en.wikipedia.org/wiki/Simplified_molecular-input_line-entry_system">Wikipedia</a>.</p>
</section>
<section id="tls-files">
<h4 id="tls-files">TLS files<a class="headerlink" href="#tls-files" title="Permalink to this heading">¶</a></h4>
<p><a class="reference external" href="http://ccp4wiki.org/~ccp4wiki/wiki/images/f/fa/Winn-Oulu08-tls.pdf">TLS
refinement</a>
attempts to model the anisotropy in the motion of a macromolecular
structure by modeling the translation,libation and screw motion of
different domains of the structure. The TLS file input to the
<em>Refinement</em> task specifies the different domains of the structure and
the initial translation,libation and screw motion tensors. For more
detailed information see
<a class="reference external" href="http://www.ccp4.ac.uk/martyn/tls_research.html">here.</a></p>
</section>
</section>
</section>
</section>


          </article>
        </div>
      </div>
    </main>
  </div>
  <footer class="md-footer">
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
          
          
        </a>
        
      </nav>
    </div>
    <div class="md-footer-meta md-typeset">
      <div class="md-footer-meta__inner md-grid">
        <div class="md-footer-copyright">
          <div class="md-footer-copyright__highlight">
              &#169; Copyright 2020, CCP4.
              
          </div>
            Created using
            <a href="http://www.sphinx-doc.org/">Sphinx</a> 5.2.2.
             and
            <a href="https://github.com/bashtage/sphinx-material/">Material for
              Sphinx</a>
        </div>
      </div>
    </div>
  </footer>
  <script src="../_static/javascripts/application.js"></script>
  <script>app.initialize({version: "1.0.4", url: {base: ".."}})</script>
  </body>
</html>