<!DOCTYPE html>
<head>
	<meta charset="utf-8">
<title>Creating Molecular Replacement Search Ensembles with CCP4MG/MrBUMP in CCP4i2</title>
	<link rel="stylesheet" href="../ccp4i2.css">
        <style>
h2,h1 {
    color:#0077aa;
    font-family: "Helvetica Neue", Helvetica, "Liberation Sans", Arial, Sans-Serif;
}
p {
    font-size: 18px;
}
li {
    font-size: 16px;
}
        </style>
</head>
<body>
<h1>Creating Molecular Replacement Search Ensembles with CCP4MG/MrBUMP in CCP4i2</h1>
<h2>Introduction</h2>
<p>
This tutorial shows how to interactively use the model preparation steps of the MrBUMP molecular replacement pipeline
using CCP4MG. MrBUMP requires only a sequence to prepare a search ensemble and therefore a sequence is the only
input required by CCP4i2 when using the CCP4MG/MrBUMP task.
</p>
<p>
The task works in the following way:
</p>
<ol>
<li>CCP4MG is opened with the specified sequence as input. CCP4MG displays this sequence.
<li>CCP4MG immediately starts running the model preparation steps of MrBUMP, which does the following:<br/>
The program <em>phmmer</em> is used to find a set of structures with similar sequences to the input, but which have no
more than specific value of sequence similarity with each other.<br/>
The structures are then pruned using <em>Scupltor</em> and aligned using <em>Gesamt</em>.
<li>The pruned and aligned structures are then loaded into CCP4MG as ribbon models.
<li>The user may then change which atoms are displayed in CCP4MG. The displayed atoms can then be saved in CCP4i2 as a
molecular replacent ensemble. Any number of different ensembles may be saved.
</ol>

<h2>Example <em>gamma</em> test data</h2>
<ol>
<li>Create a new project named e.g. <em>"mrbump_gamma"</em>
<li>Copy the CCP4 <em>gamma</em> test data to this project with the <em>Utilities -&gt; Copy demo data to project -&gt; Gamma</em>
menu item.<br/>
This will place the test data in <em>$HOME/CCP4I2_PROJECTS/mrbump_gamma/gamma</em> (Mac and Linux) or
<em>C:\Users\&lt;your username&gt;\CCP4I2_PROJECTS\mrbump_gamma\gamma</em> (Windows).
<li>Click the task menu button in CCP4i2.<br/>
Expand the <b><em>Bioinformatics including model preparation for Molecular Replacement</em></b> section.<br/>
Create a new <b><em>Interactive model preparation - CCP4MG and MrBUMP</em></b> task.
<li>As stated above, the only required input is a sequence. However, this task - like many others - uses the CCP4i2
concept of <em>AU contents</em> as its input. (In fact, CCP4i2 in the end just passes a single sequence to CCP4MG,
but this detail is not really important).<br/><br/>
To input a sequence:
<ul>
<li>Click on the <b><em>Specify AU contents</em></b> button.<br/>
Click on <b><em>Browse for sequence file button</em></b>, load <b><em>gamma.pir</em></b><br/>
Click the <b><em>Save</em></b> button.
<li>(Alternatively the drop-down menu may be used to select a previously defined AU contents.
This is relevant in real-life but not for the purposes of this tutorial.)
<li>There are two options in the CCP4MG/MrBUMP task: <b><em>Non-redundancy level for homologue search</em></b>, 
this is the value which determines how similar in sequence the search models can be; and <b><em>Cutoff threshold for phmmer results</em></b>,
this number is a score threshold specific to the homelogue search - it may be reduced in cases where there a very few homologues.
We should leave both these at their default values of 95 and 20 for this example.
</ul>
<li>Click CCP4i2's <em>Run</em> button. CCP4MG will open with the <em>gamma</em> sequence displayed. Another window
opens which shows, for the moment, the MrBUMP log file. After a while this will change to a view showing how the sequences
of the search models match the input sequence. They are ordered by <em>phmmer</em> score with the best at the top.
<li>Move the MrBUMP results window so that you can see it and the main CCP4MG window clearly. Move the <em>Show residues with
gesamt variance</em> slider to the left and see what happens to the displayed atoms. The more the slider is moved to the
left, the fewer the displayed atoms. Lowering the slider value requests CCP4MG to display atoms which are more conserved
in 3D position between all the structures. Moving the slider to the left will remove features such as loops with high 
positional variability. The far right value of 110 means show everything.
<li>Set the slider to approximately 3/4 the way to the right (about 75). Click on the <em>File</em> and select
<b><em>Save all visible to CC4i2 database</em></b>. Repeat this with the slider about 1/2 way (50).
<li>Quit CCP4MG. The CCP4i2 report for the task will show the search hits found by MrBUMP and pictures of any ensembles saved
by CCP4MG. These ensembles will also be listed in the <em>Output data</em> section of the report.
<li>Try to perform molecular replacement with your first ensemble and the <em>gamma_native.mtz</em> data.
</ol>

<h2>Example 2 <em>5yca</em> (Optional, only do if time permits)</h2>
<ol>
<li>Run an Import sequence job with the following in the <b>Or enter the rawtext of the sequence.. box</b>
<pre>
&gt;5yca 5yca
GPETHINLKVSDGSSEIFFKIKKTTPLRRLMEAFAKRQGKEMDSLRFLYDGIRIEADQTPEDLDMEDNDIIEAHRSLPAE
RNPLYKDDTLDHTPLIPKCRAQVIEFPDGPATFVRLKCTNPESKVPHFLMRMAKDSSISATSMFRSAFPKATQEEEDLEM
RWIRDNLNPIEDKRVAGLWVPPADALALAKDYSMTPFINALLEASS
</pre>

<li>Create a new CCP4MG/MrBump task.Set the AU contents to the input from job 1. Set the Non-redundancy level for homologue search to 100

<li>Run the MG/MrBump task. This time when the MrBUMP completes you willnotice that only one model is loaded. This example requires more careful consideration than the previous one. In this case their are 2 domains identified by MrBUMP. If you look at the MrBUMP results window you willsee that the white bar at the top has "Domain 1" and "Domain 2" written init and in the structures listed below, there is one which overlaps stronglywith Domain 1 and several which overlap with Domain 2. We are in fact interested in Domain 2.3. 
Uncheck "Show MrBump models", all models should disappear.Change the drop down menu from "Domain 1" to "Domain 2".Re-check "Show MrBump models", many models should appear.

<li>Set the slider to approximately 3/4 the way to the right (about 75). Click onthe File and select Save all visible to CC4i2 database. Quit CCP4MG.

<li>The output file of this task can now be used as a molecular replacement search model. This will not be covered in this tutorial
</ol>
</body>
